src/commands/externalDiff.ts

```diff
@@ -0,0 +1,87 @@
+'use strict';
+import { SourceControlResourceState, Uri, window } from 'vscode';
+import { Command, Commands } from './common';
+import { CommandContext } from '../commands';
+import { GitService } from '../gitService';
+import { Logger } from '../logger';
+import { Messages } from '../messages';
+
+enum ResourceGroupType {
+    Merge,
+    Index,
+    WorkingTree
+}
+
+interface Resource extends SourceControlResourceState {
+    readonly resourceGroupType: ResourceGroupType;
+}
+
+class ExternalDiffFile {
+    constructor(public uri: Uri, public staged: boolean) {
+    }
+}
+
+export class ExternalDiffCommand extends Command {
+
+    constructor(private git: GitService) {
+        super(Commands.ExternalDiff);
+    }
+
+    protected async preExecute(context: CommandContext): Promise<any> {
+        const repoPath = await this.git.getRepoPathFromUri(undefined);
```

Comment:
Can we move the repoPath work into execute? I try to keep the `preExecute` as just massaging context specific data into the proper args for `execute`